Fix MPR#7704

garrigue opened the pull request:
Fix [MPR#7704](https://caml.inria.fr/mantis/view.php?id=7704) by disabling unification for `AnyExtraTag` in check mode (i.e. calling `type_pat` from `Parmatch`)

Note that one could have used `AnyExtraTag` in her own program, but then it only means that in some very special case we would have a spurious warning, which is still sound.

garrigue committed:
Fix MPR#7704

garrigue committed:
add GPR number and check argument is None

trefis commented:
Couldn't you use an arbitrary string that wouldn't accepted by the parser instead of `AnyExtraTag`?
Currently (i.e. without your present patch) we can already do:
```
# function (`AnyExtraTag,_) -> () | (_,`AnyExtraTag) -> ();;
Warning 8: this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
(`AnyExtraTag, `AnyExtraTag)
- : [> `AnyExtraTag ] * [> `AnyExtraTag ] -> unit = <fun>
```
Here the warning is not spurious in the sense that there are some unmatched value, but it's also wrong, as *`AnyExtraTag* is actually matched.

If you're worried about warning 8 printing an example which doesn't look like valid ocaml then I'm guessing it will be fairly easy to update the printer to do what your doing in typecore, i.e. "is the label my globally defined `any_extra_tag`? if yes, do something smart" (`any_extra_tag` would have to move from `Parmatch` to `Printpat`, or perhaps better yet in `Typedtree`).

trefis commented:
Apart from the previous comment the fix makes sense to me.

trefis commented on Changes:
```diff
@@ -204,6 +204,9 @@ Working version
   to ill-typed columns
   (Thomas Refis, with help from Gabriel Scherer and Luc Maranget)
 
+- MPR#7704, GPR#1559: Soundness issue with private rows and module inclusion
```
I would remove the "and module inclusion" part here.

garrigue commented:
I could certainly change the string, and this is the reason I abstracted it as any_extra_tag, but I refrained to do that because it could break other tests. (I was first using AnyExtraTag*)
By the way, I don’t think it’s so easy to find a counter example: you need to find something that passes type checking, and yet is not admissible, knowing that some GADT processing is disabled in presence of polymorphic variants...

trefis commented:
> I refrained to do that because it could break other tests. (I was first using AnyExtraTag*)

I would hope it does!
(that is: I don't think that's a problem)

> By the way, I don’t think it’s so easy to find a counter example: you need to find something that passes type checking, and yet is not admissible, knowing that some GADT processing is disabled in presence of polymorphic variants...

I'm not sure what you mean... Counter example to what? What do you mean by admissible? The way I currently interpret it is "it's not easy to produce a well-typed value which isn't matched by a pattern matching" but I'm not sure that's what you meant...

I'm also not sure whether that was a general remark or an answer to my comments?
If the later, then again, I didn't mean to change anything about how you produce and typecheck such values for polymorphic variants compared to your PR. Simply to change the string so it's not a tag the user could actually write himself in the sources.
Note that a ppx might still produce that string but I find it less critical (we can blame the ppx in that case from producing code that wouldn't parse, while it's harder to blame a user for writing code that parses and typechecks).

garrigue commented:
I meant that even if you use AnyExtraTag in your pattern-matching, it is difficult to generate a false positive for the exhaustiveness check.
But maybe your intent was to change the string, independently of this bug report, to avoid confusion.
Then using something as AnyExtraTag* or « Any Extra Tag » (with spaces) would be ok: there is no need for this string to be parsable.

gasche commented:
In other parts of the compiler we have used `*foo*` for unparseable idents, so I think that `"*any extra tag*"` would be nice.

trefis commented:
> But maybe your intent was to change the string, independently of this bug report, to avoid confusion.

Yes :)

> there is no need for this string to be parsable.

Yep.

> In other parts of the compiler we have used *foo* for unparseable idents, so I think that "*any extra tag*" would be nice.

Sounds good to me.

garrigue committed:
change AnyExtraTag to *any extra tag* + log

garrigue commented:
Modified as suggested. Every body happy with that?
We also need to cherry-pick it, since this is a soundness issue.

gasche commented:
I apologize for the slowness, but I had several questions/comments while looking at this:

0. You could edit the Changes to add Thomas Refis as a reviewer.

1. I don't really like the way the warning looks like today when it mentions this magic constructor. Currently `` `*any extra tag*`` is printed. The `` ` `` symbol (note: to write them in Markdown syntax, I use double-backquote for code quoting, with a space around the single-backquote) is out of place, and I think that "any other tag" would read better than "any extra tag", and for the user I think that `<any other tag>` would read better than `*any extra tag*` -- but still in the codebase I would support the use of star by convention. Could we do as @trefis suggested and special-case this constructor value in the printer, so that we can evolve the user-displayed string without having to also change the typechecker implementation?

2. If you intend to have this in the 4.06 stable branch, I thought that I could have a look at whether the fix itself is correct. I had a look and my answer is that there is not enough information in your patch or comments (here and on the PR) to understand what is happening here. Would you care to explain why disabling unification of this `any_extra_tag` constructor is the right thing to do to fix this soundness issue?

Below is a summary of what I understand so far.

The `any other tag` special tag is produced by `Parmatch.exhaust`, whose (undescribed) specification is: produce a value not matched by its input matrix `pss`. `any other tag` is produced when the discriminators of a column are polymorphic variant constructors. (We could instead generate a valid tag that is not among those head constructors, by searching from A to Z, and then adding numer suffixes, not unlike what `build_other_constant` is doing.)

If I follow correctly, the (only) way this can reach the type-checker is by `Parmatch.check_partial`, which gets called by `Typecore.check_partial` (matching.ml has a different function of the same name, just for fun). `check_partial` type-checks the resulting pattern (all counter-examples have been or-ified together) and declares the matching (or let-binding) partial only if the type-checking fails.

Now comes the question: the idea of generating a value that contains a "other" tag not listed in the pattern-matching row seems perfectly sensible to me, but why does it need to be special-cased in the type checker for soundness? Your fix disables unification as follows:

```diff
-      unify_pat_types loc !env (newty (Tvariant row)) expected_ty;
+      if constrs = None || l <> Parmatch.any_extra_tag || sarg <> None then
+        unify_pat_types loc !env (newty (Tvariant row)) expected_ty;
```

What are the three conditions for? Why does `any_extra_tag` need to be special-cased? How does that fix the bug (i suppose it makes the counter-example type-check again?)? Could you maybe add a comment in the code to point to the MPR# to explain this non-intuitive test?

Would it be possible to add your reproduction test to the testsuite?

garrigue committed:
add Thomas Refis as reviewer

garrigue committed:
add comment in code

gasche commented:
(If you want to merge this in the 4.06 maintenance branch, which does make sense given that it is a soundness bug with a small fix, you should move the Changes entry to the relevant section.)

garrigue commented:
0. Done.
1. Honestly, I don't think it is worth it. We are just talking about the wording of a warning. I'm ready to use any string you please for the tag, but please let us keep it static, and not add any code that would require debugging.
2. `any_extra_tag` is a special tag that is introduced during exhaustiveness checking to denote the tags that could be hidden by an abstract row variable. By definition, no tag can be unified with an abstract row variable, so unification has to be disabled when seeing it. The extra check that `sarg = None` is there to ensure that we are not losing anything more by not doing the unification (it is not strictly necessary if `any_extra_tag` is really a special tag that cannot appear in concrete code, but this avoids depending on that property).

And of course the reproduction test is there from the beginning.

garrigue committed:
move log to 4.06 maintenance section

gasche commented:
I see, thanks for the explanation -- I had ignored the "private row type" aspect.

trefis commented on typing/typecore.ml:
```diff
@@ -1249,7 +1249,10 @@ and type_pat_aux ~constrs ~labels ~no_existentials ~mode ~explode ~env
+      (* PR#7404: allow any_extra_tag blindly, as it would not unify with
+         the abstract row variable *)
+      if constrs = None || l <> Parmatch.any_extra_tag || sarg <> None then
+        unify_pat_types loc !env (newty (Tvariant row)) expected_ty;
```
I'd be tempted to rewrite this to:
```ocaml
if l <> Parmatch.any_extra_tag then
  unify_pat_types loc !env (newty (Tvariant row)) expected_ty
else begin
  assert (constrs <> None);
  assert (sargs = None)
end;
```
but apparently you don't want to enforce that `any_extra_tag` cannot appear in the input and be produced only by parmatch...

I feel like it would be better to enforce it, but unless someone else shares that opinion I'm fine with merging your current implementation.

trefis reviewed (approved):
Apart from the small rewrite I am proposing, which I do not feel strongly about, the only remaining topic of discussion seems to be about how to pretty-print this newly introduced `Parmatch.any_extra_tag`.
Which, I believe, could be discussed independently from the current PR.

So I'd be OK with this PR being merged as is.

garrigue committed:
change name to <some other tag>

garrigue commented:
I added the assertion, and changed the name to `<some other tag>`, which looks nicer in the warning.

Note that one could check `row_more` in `Parmatch` to generate different names depending on whether there is an abstract row or not, and even as you suggested use real tag names when the row variable is a normal one (checking they are not in `env`). But since the goal of this PR is to recover soundness, and this could potentially make the fix more fragile, I'll leave that for later.
